### ISSUE TITLE
Question to Training routine for adversarial loss

### ISSUE THREAD

**Issue by nkoser** [Lornatang/SRGAN-PyTorch#42]
Hi, 

I have a question regarding the training of the GAN and namely I wonder why you don't call the optimizer from the generator as well as from the discriminator with zero_grad() at the beginning of the for loop and only call discriminator.zero_grad() and generator.zero_grad() respectively ? Unfortunately I'm still quite new to Pytorch and have always used Tensorflow before and could imagine that it amounts to the same thing ? 

Further, I also wonder why, for example, you don't do something like G.train(False) and D.train(True) when updating the discriminator and vice versa ? 

I hope you can answer my questions and I remain with kind regards
Niklas

**Comment by Lornatang** [comment]
1. You need to clear the gradient information of the previous iteration during each training, so use zero_grad().
2. Because the .detach() function is already used, there is no need to set the .train(False) variable.

**Comment by nkoser** [comment]
Hi Lornatang, 
thanks for the information

1.  I just saw that over all always so far that deleting the gradients went via optimzer.zero_grad() and not via model.zero_grad(). So I was a bit confused. So can I assume that optimzer.zero_grad() and model.zeor_grad() have the same effect ? 

2. Yes you're right

**Comment by Lornatang** [comment]
If you only have one model, they are equivalent

**Comment by nkoser** [comment]
okay thanks :D, but here I have two models or ?

**Comment by Lornatang** [comment]
They are separate

**Comment by nkoser** [comment]
Exactly, that's what I thought. Again another question: Is d_loss_sr -> backward(), d_loss_hr -> backward() equivalent to d_loss = d_loss_hr +d_loss_sr -> d_loss.backward

**Comment by Lornatang** [comment]
No.~ See GAN paper

**Comment by nkoser** [comment]
Thank you very much for your quick help
